Adding metrics to Swift Jobs

thoven87 opened the pull request:
Here's my first attempt at adding Observability to the Swift jobs project.

I would love some feedback on how to make metrics better.  cc @adam-fowler 

Since everyone might use different tools to display their dashboards, I will attempt to provide a grafana dashboard as it is what I am most familiar with.

Lastly, most of the tests are a compilation of the current job tests and [Hummingbird](https://github.com/hummingbird-project/hummingbird/blob/main/Tests/HummingbirdTests/MetricsTests.swift) metrics tests.

adam-fowler commented on Sources/Jobs/JobQueue.swift:
```diff
@@ -41,11 +42,17 @@ public struct JobQueue<Queue: JobQueueDriver>: Service {
         let buffer = try JSONEncoder().encodeAsByteBuffer(jobRequest, allocator: self.allocator)
         let id = try await self.queue.push(buffer)
-        self.handler.logger.debug("Pushed Job", metadata: ["_job_id": .stringConvertible(id), "_job_type": .string(jobRequest.id.name)])
+        Counter(label: "queued_jobs_counter", dimensions: [("queue_name", jobRequest.id.name)]).increment()
```
queue_name is a little misleading, shouldn't this be job_name, or JobName to coincide with the log

adam-fowler commented on Sources/Jobs/JobQueueHandler.swift:
```diff
@@ -119,4 +126,40 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
+        ).recordNanoseconds(DispatchTime.now().uptimeNanoseconds - startTime)
+
+        // Increment job counter base on status
+        if error != nil {
```
if let error

adam-fowler commented on Sources/Jobs/JobQueueHandler.swift:
```diff
@@ -119,4 +126,40 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
+            if error is CancellationError {
+                Counter(
+                    label: "cancelled_jobs_counter",
+                    dimensions: [("queue_name", name)]
```
Again queue_name, should this not be JobName

adam-fowler commented on Sources/Jobs/JobQueueHandler.swift:
```diff
@@ -119,4 +126,40 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
+    ) {
+        // Calculate job execution time
+        Timer(
+            label: "\(name)_job_duration",
```
Shouldn't the name just be the a dimension. It is elsewhere

adam-fowler commented on Tests/JobsTests/MetricsTests.swift:
```diff
@@ -0,0 +1,294 @@
+import ServiceLifecycle
+import XCTest
+
+final class TestMetrics: MetricsFactory {
```
I've just made some changes to this to make it compile in Swift 6 mode. You'll need to wrap all this mutable state in NIOLockedValueBoxes. If you just copied and pasted this from Hummingbird, you can probably do the same again. Although you will need to edit the accesses in the `XCTAssert...` calls

thoven87 replied:
> I've just made some changes to this to make it compile in Swift 6 mode. You'll need to wrap all this mutable state in NIOLockedValueBoxes. If you just copied and pasted this from Hummingbird, you can probably do the same again. Although you will need to edit the accesses in the `XCTAssert...` calls

Thanks Adam! I have updated the helpers with the help of your previous PR.

adam-fowler commented on Package.swift:
```diff
@@ -28,6 +29,7 @@ let package = Package(
                 .product(name: "NIOCore", package: "swift-nio"),
                 .product(name: "NIOFoundationCompat", package: "swift-nio"),
                 .product(name: "ServiceLifecycle", package: "swift-service-lifecycle"),
+                .product(name: "Metrics", package: "swift-metrics"),
```
Minor nit, I added these in alphabetical order

adam-fowler reviewed (commented):
Here are some initial comments. In general this look good, I was wondering if it is worthwhile adding a metric for the retry count.

For another PR if we added a dateCreated to the QueuedJob, we could add a time job queued metric. In a way dateCreated is probably a useful piece data regardless.

thoven87 commented:
> Here are some initial comments. In general this look good, I was wondering if it is worthwhile adding a metric for the retry count.
> 
> For another PR if we added a dateCreated to the QueuedJob, we could add a time job queued metric. In a way dateCreated is probably a useful piece data regardless.

Thanks, for the feedback. I added a retry counter. I will open another PR to add both `createdAt` and maybe Job name? Right now, `Job.id.name` does not always exist in every context. In some context, it's just Job.id which is a UUID.

Joannis commented on Sources/Jobs/JobQueueHandler.swift:
```diff
@@ -64,7 +67,8 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
         var logger = logger
-        logger[metadataKey: "_job_id"] = .stringConvertible(queuedJob.id)
+        let startTime = DispatchTime.now().uptimeNanoseconds
+        logger[metadataKey: "JobId"] = .stringConvertible(queuedJob.id)
```
Here you're making each word Uppercased, whereas the other key is snake_cased. I'd prefer sticking to one.

thoven87 replied:
> Here you're making each word Uppercased, whereas the other key is snake_cased. I'd prefer sticking to one.

Agreed! Fixed in the latest commit.

Joannis commented:
Looks great! Just one bit of consistency

thoven87 commented:
I think this Pull request is ready. Below is a dashboard I created to test this implementation.

Next step would be to add tracing and queued at to compute avg wait time.
<img width="1989" alt="Screenshot 2024-08-18 at 12 35 45 PM" src="https://github.com/user-attachments/assets/b6d953fa-ce30-43c6-8965-cae807456ed2">

<img width="1985" alt="Screenshot 2024-08-18 at 12 50 28 PM" src="https://github.com/user-attachments/assets/4cc1cf4d-2aa6-4742-81a5-9ca12d8ceadb">

thoven87 commented:
also previous implementation of the dashboard to show the failed metrics
<img width="1984" alt="Screenshot 2024-08-17 at 8 46 21 PM" src="https://github.com/user-attachments/assets/52df52e6-b349-4582-b751-6f5ed66137e7">

adam-fowler commented on Sources/Jobs/JobQueueHandler.swift:
```diff
@@ -119,4 +130,70 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
+        self.updateJobMeters()
+    }
+    
+    private func updateJobMeters() {
```
Can we rename this. It isn't clear what it is doing. Also if this is called if the job fails to decode, or it couldn't find the job id then processing is decremented when it hasn't been incremented.

I'm wondering if it be easier to break out the queued and processing counter editing from here and just place at the top of `runJob`. You'd also have to remove the processing `decrement` when you retry a job. 

```swift
Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
defer {
    Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
}

thoven87 replied:
> Can we rename this. It isn't clear what it is doing. Also if this is called when the job fails to decode, or it couldn't find the job id then processing is decremented when it hasn't been incremented.
> 
> I'm wondering if it be easier to break out the queued and processing counter editing from here and just place at the top of `runJob`. You'd also have to remove the processing `decrement` when you retry a job.
> 
> ```swift
> Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
> Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
> defer {
>     Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
> }
> ```

This is much cleaner! I will update accordingly, thanks!

thoven87 replied:
@adam-fowler I updated the PR with the suggestion above.

thoven87 commented:
> This is great

Please merge at your convenience! What's the best place to share RED metrics dashboards? I am in the process of creating a few  Grafana dashboards for Hummingbird.

There would be three dashboards in total: 

1 -  Display the overall health of an application including jobs and http metrics, spans and error logs. 
2 - Display Http metrics, spans and error logs
3 - Display Jobs metrics, spans and error logs

adam-fowler commented:
Just waiting on a final approval from @Joannis

Joannis reviewed (approved):
Looks awesome!

Joannis commented:
@thoven87 I'd love to post them either in a docs section, on docs.hummingbird.codes as a DocC guide. I think we can also highlight it on swiftonserver.com

thoven87 commented:
> @thoven87 I'd love to post them either in a docs section, on docs.hummingbird.codes as a DocC guide. I think we can also highlight it on swiftonserver.com

@Joannis I suppose for DooC guide it would need to be added in a section in [this repository](https://github.com/hummingbird-project/hummingbird-docs) Indeed, it would be great to highlight it on swiftonserver.com as well.

thoven87 commented:
> You could possibly include something in the observability docs but https://docs.hummingbird.codes/2.0/documentation/hummingbird/loggingmetricsandtracing.
> 
> I think an end to end metrics and tracing article in swiftonserver.com which uses these would be a great addition.

Thanks Adam, I will take a look and see what I can do.
